content/articles/socket-programming-in-the-dotnet-framework-using-csharp/index.md

```diff
@@ -0,0 +1,230 @@
+                    {
+                        Console.WriteLine($"Read:{content.Length} data from \n socket data:{content}");
+                        Send(handler, content);
+                    }
+                }
+                else
+                {
+                    handler.BeginReceive(state.data, 0, ProgramState.dataSize, 0, new AsyncCallback(ReadCallBack), state);
+                }
+            }
+```
+
+When we create the method above, we get an error that requires us to create the `Send` method. In this method, we will have a byte array and name it `sizeOfData` and pass in the content of our data. 
+
+We then use the `BeginSend` function on our handler and pass in the `sizeOfData`, data length, handler as parameters. `new AsyncCallBack` is a parameter in this function that receives `SendCallBack` as another parameter.
+
+```C#
+private static void Send(Socket handler, string content)
+            {
+                byte[] sizeOfData = Encoding.ASCII.GetBytes(content);
+                handler.BeginSend(sizeOfData, 0, sizeOfData.Length, 0, new AsyncCallback(SendCallBack), handler);
+            }
+```
+
+The `SendCallBack` argument in the above method generates an error that requires us to create another method called `SendCallBack`. In this method, we will use a `try...catch` method, add a socket handler, and define the data sent to the client. 
+
+We will be closing the connections on both ends once data is sent. The `catch `method is used to check for any errors in the connection.
+
+```C#
+private static void SendCallBack(IAsyncResult ar)
+            {
+                try
+                {
+                    Socket handler = (Socket)ar.AsyncState;
+                    int dataSent = handler.EndSend(ar);
+                    Console.WriteLine($"Send:{dataSent} to client");
+                    handler.Shutdown(SocketShutdown.Both);
+                    handler.Close();
+                }
+                catch (Exception e)
+                {
+                }
+            }
+```
+
+Now the last thing is to implement our application in the `main` function. We will use the `Console.WriteLine` method to display a message on the screen. 
+
+The `Console.ReadLine` will be used to take data from the user and then we will start our listener in the main function.
+
+```C#
+        static void Main(string[] args)
+        {
+            Console.WriteLine("Press enter key to continue...");
+            Console.ReadLine();
+            AsyncSocketListener.StartListener();
+            Console.ReadLine();
+        }
+```
+
+### Conclusion
+From this article, we can deduce socket programming is a better way for devices to communicate on a network. For example, if a computer device sends data on port `4444` and uses the TCP protocol, the data can be received on our device. This is because our device will be acting as a server and the sender will be acting as a server.
```

Comment:
```suggestion
From this article, we can deduce that socket programming is a better way for devices to communicate over a network. For example, if a device sends data on port `4444`  using the TCP protocol, the data can be received the other end. This is because our device will be acting as a recipient and the sender will be acting as a server.